docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,431 @@
+    cluster.
+
+In the above scheme, force deleting a pod releases the lock on that pod and
+allows higher level components to proceed to create a replacement.
+
+It has been requested that force deletion be restricted to privileged users.
+That limits the application owner in resolving partitions when the consequences
+of force deletion are understood, and not all application owners will be
+privileged users. For example, a user may be running a 3 node etcd cluster in a
+pet set. If pet 2 becomes partitioned, the user can instruct etcd to remove
+pet 2 from the cluster (via direct etcd membership calls), and because a quorum
+exists pets 0 and 1 can safely accept that action. The user can then force
+delete pet 2 and the pet set controller will be able to recreate that pet on
+another node and have it join the cluster safely (pets 0 and 1 constitute a
+quorum for membership change).
+
+This proposal does not alter the behavior of finalizers - instead, it makes
+finalizers unnecessary for common application cases (because the cluster only
+deletes pods when safe).
+
+### Fencing
+
+The changes above allow Pet Sets to ensure at-most-one pod, but provide no
+recourse for the automatic resolution of cluster partitions during normal
+operation. For that, we propose a **fencing controller** which exists above
+the current controller plane and is capable of detecting and automatically
+resolving partitions. The fencing controller is an agent empowered to make
+similar decisions as a human administrator would make to resolve partitions,
+and to take corresponding steps to prevent a dead machine from coming back
+to life automatically.
+
+Fencing controllers most benefit services that are not innately replicated
+by reducing the amount of time it takes to detect a failure of a node or
+process, isolate that node or process so it cannot initiate or receive
+communication from clients, and then spawn another process. It is expected
+that many PetSets of size 1 would prefer to be fenced, given that most
+applications in the real world of size 1 have no other alternative for HA
+except reducing mean-time-to-recovery.
+
+While the methods and algorithms may vary, the basic pattern would be:
+
+1. Detect a partitioned pod or node via the Kubernetes API or via external
+   means.
+2. Decide whether the partition justifies fencing based on priority, policy, or
+   service availability requirements.
+3. Fence the node or any connected storage using appropriate mechanisms.
+
+For this proposal we only describe the general shape of detection and how existing
+Kubernetes components can be leveraged for policy, while the exact implementation
+and mechanisms for fencing are left to a future proposal. The fencing controller
+would be able to leverage a number of systems including but not limited to:
+
+* Cloud control plane APIs such as machine force shutdown
+* Additional agents running on each host to force kill process or trigger reboots
```

Comment:
why is this necessary? Is kubelet is robust enough, this should not be necessary and we do want kubelet to be robust in general